### DISCUSSION TITLE
'emcc: DEBUG: using binaryen' delayed when  ASYNCIFY is turn on

### DISCUSSION THREAD

**Discussion by aharondavid** [emscripten-core/emscripten#16959] [Q&A]
Hi,
After turn on the EMCC_DEBUG, I found that the build are delayed in `emcc:DEBUG: using binaryen`
Is there an option to decrease the build time when linking with ASYNCIFY in debug (-g) ?
otherwise, it make it unusable ...

**Comment by kripken** [comment]
You can try to optimize the code more in LLVM before it gets to binaryen, like compiling to object files with `-Os`. Less code reaching binaryen would be faster.

Otherwise, profiling your build might find it's hitting something we can optimize, if you see something very slow.

Definitely worth optimizing stuff we can, but at least in the long term this will go away entirely once wasm VMs have stack switching, as then we don't need to run the Asyncify transform (https://github.com/emscripten-core/emscripten/pull/16779).

**Comment by aharondavid** [comment]
unfortunately, compiling with -Os doesn't reduced the build time, here is the emmc debug profiler output:
(executed on some sample code, the real project is match bigger and the build time (linking) there, with ASYNCIFY is more than 15 minute!. without ASYNCIFY is less than minute).
the major consuming time :
**profiler:DEBUG: block "get_metadata_binaryen" took 14.937 seconds
profiler:DEBUG: block "emscript" took 15.245 seconds
profiler:DEBUG: block "post_link" took 67.944 seconds**

According to the debug log below, Is there any additional  operations that we can do for decrease that time?

full log:

```
profiler:DEBUG: block "get_metadata_binaryen" took 14.937 seconds
building:DEBUG: saving debug copy C:\Users\ADMINI~1\AppData\Local\Temp\emscripten_temp\emcc-2-post_finalize.wasm
emscripten:DEBUG: emscript: js compiler glue
shared:DEBUG: successfully executed C:/emsdk/node/14.18.2_64bit/bin/node.exe C:\emsdk\upstream\emscripten\src\compiler.js C:\Users\ADMINI~1\AppData\Local\Temp\emscripten_temp\tmp2dp4voug.json
emscripten:DEBUG:   emscript: glue took 0.18899989128112793 seconds
building:DEBUG: saving debug copy C:\Users\ADMINI~1\AppData\Local\Temp\emscripten_temp\emcc-3-original.js
profiler:DEBUG: block "emscript" took 15.245 seconds
profiler:DEBUG: block "source transforms" took 0.000 seconds
emcc:DEBUG: using binaryen
emcc : warning : running limited binaryen optimizations because DWARF info requested (or indirectly required) [-Wlimited-postlink-optimizations]
shared:DEBUG: successfully executed C:/emsdk/upstream\bin\wasm-opt --post-emscripten -Os --low-memory-unused --asyncify --pass-arg=asyncify-imports@env.invoke_*,env.__call_main,env.emscripten_sleep,env.emscripten_wget,env.emscripten_wget_data,env.emscripten_idb_load,env.emscripten_idb_store,env.emscripten_idb_delete,env.emscripten_idb_exists,env.emscripten_idb_load_blob,env.emscripten_idb_store_blob,env.SDL_Delay,env.emscripten_scan_registers,env.emscripten_lazy_load_code,env.emscripten_fiber_swap,wasi_snapshot_preview1.fd_sync,env.__wasi_fd_sync,env._emval_await,env._dlopen_js,env.__asyncjs__* --zero-filled-memory --strip-producers C:/Projects/simple_opengl-master/Emscripten/Debug/TheMainProject.wasm -o C:/Projects/simple_opengl-master/Emscripten/Debug/TheMainProject.wasm -g --mvp-features --enable-threads --enable-mutable-globals --enable-bulk-memory --enable-sign-ext
building:DEBUG: saving debug copy C:\Users\ADMINI~1\AppData\Local\Temp\emscripten_temp\emcc-4-wasm-opt.wasm
building:DEBUG: saving debug copy C:\Users\ADMINI~1\AppData\Local\Temp\emscripten_temp\emcc-5-byn.wasm
building:DEBUG: supporting wasm memory growth with pthreads
shared:DEBUG: successfully executed C:/emsdk/node/14.18.2_64bit/bin/node.exe C:\emsdk\upstream\emscripten\tools\acorn-optimizer.js C:\Users\ADMINI~1\AppData\Local\Temp\emscripten_temp\TheMainProject.js growableHeap
building:DEBUG: saving debug copy C:\Users\ADMINI~1\AppData\Local\Temp\emscripten_temp\emcc-6-growableHeap.js
profiler:DEBUG: block "binaryen" took 52.551 seconds
shared:DEBUG: successfully executed C:/emsdk/node/14.18.2_64bit/bin/node.exe C:\emsdk\upstream\emscripten\tools\preprocessor.js C:\Users\ADMINI~1\AppData\Local\Temp\emscripten_temp\settings.js worker.js --expandMacros
emcc:DEBUG: Private module export name substitution with Module
building:DEBUG: saving debug copy C:\Users\ADMINI~1\AppData\Local\Temp\emscripten_temp\emcc-7-module_export_name_substitution.js
root:DEBUG: move: C:\Users\ADMINI~1\AppData\Local\Temp\emscripten_temp\TheMainProject.js.pgrow.js.module_export_name_substitution.js -> C:/Projects/simple_opengl-master/Emscripten/Debug/TheMainProject.js
profiler:DEBUG: block "final emitting" took 0.148 seconds
profiler:DEBUG: block "post_link" took 67.944 seconds
emcc:DEBUG: total time: 70.63 seconds
profiler:DEBUG: block "main" took 70.629 seconds
not cleaning up temp files since in debug-save mode, see them in C:\Users\ADMINI~1\AppData\Local\Temp\emscripten_temp
not cleaning up temp files since in debug-save mode, see them in C:\Users\ADMINI~1\AppData\Local\Temp\emscripten_temp
not cleaning up temp files since in debug-save mode, see them in C:\Users\ADMINI~1\AppData\Local\Temp\emscripten_temp
```

**Reply by kripken** [reply]
For a very large wasm file, there isn't a simple way to speed that up. Asyncify is a fairly complex transform that can be slow.

Profiling might find something that could be optimized, as I said before. You can try to build with `BINARYEN_PASS_DEBUG=1` in the env, which will time passes. Perhaps one of the internal parts of Asyncify is particularly slow.

**Comment by aharondavid** [comment]
hello @kripken.
sorry for the delay in responding .
I tried the BINARYEN_PASS_DEBUG=1 but it doesn't tell me something that can be solve.
and probably, as you said, this flag doesn't  fit to a big wsam.
I am looking for alternative way to do a c++ sync calls. 
I thought that an infinit emscripten_set_main_loop can be an option for that, but I didn't find a way to return from infinit emscripten_set_main_loop.
is there any other option to do a sync call ?

**Comment by aharondavid** [comment]
I had found the answer in https://github.com/emscripten-core/emscripten/issues/17130.
Thanks.
